AWS: Remove obsolete permissions on ELB security group

justinsb opened the pull request:
The ingress CIDRs are going to be dynamic, and in general we don't want
to leave old ingress rules around.

Fix #21895

Depends on #21907

justinsb committed:
AWS: Remove extra load balancer security group ingress rules

The ingress CIDRs are going to be dynamic, and in general we don't want
to leave old ingress rules around.

Fix #21895

mikedanese commented:
Ok, 21907 is lgtm'd, i plan on reviewing this tomorrow.

justinsb commented:
Thanks @mikedanese ...  #21907 got caught up in the merge problems last night (#22435).  But I'm trying to push it through...

I did rebase this on the LGTMed #21907, so hopefully 21907 will go through shortly (it just made it back onto the submit queue) and then this one should be clean.

This PR is going to just be for the final commit.

mikedanese commented on pkg/cloudprovider/providers/aws/sets_ippermissions.go:
```diff
@@ -0,0 +1,146 @@
+
+type IPPermissionSet map[string]*ec2.IpPermission
+
+func NewIPPermissionSet(items ...*ec2.IpPermission) IPPermissionSet {
```
Was this autogenerated? cc @lavalamp

justinsb replied:
It wasn't - we can't currently autogenerate things that can't be map-keys.  We had the same thing in #21907 for sets of net.IPNet.  There we ended up putting it into pkg/util/net/sets, expected alias for the import being netsets, with a readme that we will likely move it to pkg/util/sets when/if we get autogen to work.  And the API for netsets.IPNet should match the API for sets, at least as much as is implemented.

This one is AWS only, so I don't think we should let it escape pkg/cloudproviders/providers/aws.

mikedanese commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1590,10 +1591,73 @@ func isEqualUserGroupPair(l, r *ec2.UserIdGroupPair, compareGroupUserIDs bool) b
+
+		request := &ec2.RevokeSecurityGroupIngressInput{}
+		request.GroupId = &securityGroupId
+		request.IpPermissions = remove.List()
```
Considering there is a limit on number of rules per security group (100 or something) should we be regrouping?

justinsb replied:
Probably yes... but:

(1) this is only used on the ELB security group currently so it would require someone setting (ports * CIDRs) must be > 100.
(2) it isn't clear to me exactly what calls we would have to make to remove a single permission from a compound rule, and I worry about accidentally opening more than intended if I get it wrong.

How about I add a TODO here to address this properly in future (maybe we see if anyone hits the issue!)

mikedanese commented:
IIUC this makes it so people can't add their own rules to a security group. I think theirs a use case for being able to add your own rules to the k8s security group, and also seems like it might break someone if they upgrade and had been using manually created rules.

mikedanese commented:
Ahh diffent security groups for elbs than nodes so last point probably doesn't make sense right?

justinsb commented:
> IIUC this makes it so people can't add their own rules to a security group

Good observation.  Yes, you can't (or you can't expect them to survive).

> I think theirs a use case for being able to add your own rules to the k8s security group,

The way around this that I've been pushing is to have two security groups on your ELBs/instances.  The one that is tagged with the KubernetesCluster tag is owned by kubernetes, and abandon hope all you who enter there.  But k8s will not touch the second security group, and if you want to add your own rules, that is where you would do that.  In general though, I think going behind k8s's back and doing stuff directly in the API is not something we can support generally.  You should expect that a future version of k8s will reconcile your changes to the k8s state of the world, as k8s adds more and more functionality.

We can carve out exceptions which we will allow, like "if you want to change security group rules manually on an ELB, add a second security group".
  
> and also seems like it might break someone if they upgrade and had been using manually created rules.

Yes.  But upgrades are non-existent right now anyway on AWS...  still hoping to get to it for 1.2.

justinsb commented:
> Ahh diffent security groups for elbs than nodes so last point probably doesn't make sense right?

No, I think it makes just as much sense to add a security group rule to a node as to an ELB :-)  Arguably #21907 reduces the need for the ELB case, but there's always an exception...

mikedanese commented:
> The way around this that I've been pushing is to have two security groups on your ELBs/instances.

That sounds fine

justinsb commented:
Added a comment that we should probably regroup rules in future, and rebased so there is only 1 commit now in this PR. Also #21907 has now merged so this is now mergeable if you approve @mikedanese

justinsb commented:
Fixes a P1 so marking as P1

justinsb commented:
@k8s-bot test this github flake #22270
